File: lib/data/services/server_url_normalizer.dart
/// Normalizes user-entered server URLs for Xtream-compatible APIs.
class ServerUrlNormalizer {
  static const _apiPathSuffixes = [
    'player_api.php',
    'panel_api.php',
    'get.php',
    'xmltv.php',
  ];

  /// Returns a base server URL without API paths or credential query strings.
  ///
  /// Examples:
  /// - `http://host:8080/player_api.php?user=x` -> `http://host:8080`
  /// - `cf.example.com` -> `http://cf.example.com`
  static String normalize(String raw) {
    var input = raw.trim();
    if (input.isEmpty) {
      throw ArgumentError('Server URL cannot be empty.');
    }

    if (!input.startsWith('http://') && !input.startsWith('https://')) {
      input = 'http://$input';
    }

    final uri = Uri.parse(input);
    if (uri.host.isEmpty) {
      throw ArgumentError('Server URL must include a hostname.');
    }

    final segments = uri.pathSegments.where((s) => s.isNotEmpty).toList();
    while (segments.isNotEmpty &&
        _apiPathSuffixes.contains(segments.last.toLowerCase())) {
      segments.removeLast();
    }

    final buffer = StringBuffer()
      ..write(uri.scheme)
      ..write('://')
      ..write(uri.host);

    if (uri.hasPort) {
      buffer.write(':${uri.port}');
    }

    if (segments.isNotEmpty) {
      buffer.write('/${segments.join('/')}');
    }

    var normalized = buffer.toString();
    if (normalized.endsWith('/')) {
      normalized = normalized.substring(0, normalized.length - 1);
    }

    return normalized;
  }

  /// If the pasted URL included `username` / `password` query params, return them.
  static ({String? username, String? password}) extractEmbeddedCredentials(
    String raw,
  ) {
    final trimmed = raw.trim();
    if (!trimmed.contains('?')) {
      return (username: null, password: null);
    }

    try {
      final uri = Uri.parse(
        trimmed.startsWith('http') ? trimmed : 'http://$trimmed',
      );
      return (
        username: uri.queryParameters['username'],
        password: uri.queryParameters['password'],
      );
    } catch (_) {
      return (username: null, password: null);
    }
  }
}

File: webos/src/api/server-url-normalizer.ts
const API_PATH_SUFFIXES = [
  'player_api.php',
  'panel_api.php',
  'get.php',
  'xmltv.php',
];

export function normalizeServerUrl(raw: string): string {
  let input = raw.trim();
  if (!input) {
    throw new Error('Server URL cannot be empty.');
  }

  if (!input.startsWith('http://') && !input.startsWith('https://')) {
    input = `http://${input}`;
  }

  const uri = new URL(input);
  if (!uri.hostname) {
    throw new Error('Server URL must include a hostname.');
  }

  const segments = uri.pathname.split('/').filter((s) => s.length > 0);
  while (
    segments.length > 0 &&
    API_PATH_SUFFIXES.includes(segments[segments.length - 1].toLowerCase())
  ) {
    segments.pop();
  }

  let normalized = `${uri.protocol}//${uri.hostname}`;
  if (uri.port) {
    normalized += `:${uri.port}`;
  }
  if (segments.length > 0) {
    normalized += `/${segments.join('/')}`;
  }
  if (normalized.endsWith('/')) {
    normalized = normalized.slice(0, -1);
  }
  return normalized;
}

export function extractEmbeddedCredentials(raw: string): {
  username: string | null;
  password: string | null;
} {
  const trimmed = raw.trim();
  if (!trimmed.includes('?')) {
    return { username: null, password: null };
  }
  try {
    const uri = new URL(
      trimmed.startsWith('http') ? trimmed : `http://${trimmed}`,
    );
    return {
      username: uri.searchParams.get('username'),
      password: uri.searchParams.get('password'),
    };
  } catch {
    return { username: null, password: null };
  }
}
